internal/ts_test.bzl

```diff
@@ -0,0 +1,102 @@
+load("@build_bazel_rules_nodejs//internal:node.bzl",
+    "sources_aspect",
+    "expand_path_into_runfiles",
+)
+
+_CONF_TMPL = "//internal/karma:karma.conf.js"
+_LOADER = "//internal/karma:test-main.js"
+
+def _ts_test_impl(ctx):
+  conf = ctx.actions.declare_file(
+      "%s.conf.js" % ctx.label.name,
+      sibling=ctx.outputs.executable)
+
+  files = depset(ctx.files.srcs)
+  for d in ctx.attr.deps:
+    if hasattr(d, "node_sources"):
+      files += d.node_sources
+    elif hasattr(d, "files"):
+      files += d.files
+
+  TMPL_files = "\n".join([
+      "      '%s'," % expand_path_into_runfiles(ctx, f.short_path)
+      for f in files
+  ])
+
+  # root-relative (runfiles) path to the directory containing karma.conf
+  config_segments = len(conf.short_path.split("/"))
+
+  ctx.actions.expand_template(
+      output = conf,
+      template =  ctx.file._conf_tmpl,
+      substitutions = {
+          "TMPL_runfiles_path": "/".join([".."] * config_segments),
+          "TMPL_files": TMPL_files,
+          "TMPL_workspace_name": ctx.workspace_name,
+      })
+
+  ctx.actions.write(
```

Comment:
Is the plan to have a karma_web_test_suite() or ts_web_test_suite() to run all of these Karma scripts?

Reply:
That's what I want to understand from you :)
Right now each `ts_test` just runs karma on the conf file it produces

Reply:
My original plan was that `ts_test()`s don't run by themselves because starting multiple Karma instances is expensive. I wanted to have a top-level rule/script that starts the Karma instance and shares it across all `ts_test()`s that it depends on.
But as you said "Doesn't yet include any way to discover and connect to an existing karma instance, so it starts a bunch of browser windows, one per test target. Can add that later". So I totally OK with the current way it is.